### PR TITLE
Taproot: BIP32 extended keys using Scep256k1 keys instead of bitcoin ECDSA

### PR DESCRIPTION
**Pull request by dr-orlovsky**
This is third step required to introduce Schnorr key support according to #588. This PR starts API-breaking changes and is follow-up to non-API breaking #589, which is already merged.

PR rationale: BIP32 does not support uncompressed keys and using type with compression flag was a mistake

### CODE REVIEW

File: src/util/bip32.rs
Comment (thomaseizinger) [inline src/util/bip32.rs:506]: We have a `From<secp256k1::Error> for Error` implementation so I think you should be able to get rid of the `map_err`:

```suggestion
            private_key: secp256k1::SecretKey::from_slice(&hmac_result[..32])?,
```

File: src/util/bip32.rs
Comment (thomaseizinger) [inline src/util/bip32.rs:637]: Small typo:

```suggestion
    /// Constructs ECDSA compressed public key matching internal public key representation.
```

**Review by thomaseizinger — COMMENTED** [review]
ACK d155115a37618ac4e6e2924d445a453b40ee7c17

File: src/util/bip32.rs
Comment (dr-orlovsky) [reply src/util/bip32.rs]: Fixed (in many other places as well)

**Comment by dr-orlovsky** [comment]
Rebased on master, addressed comments, now ready for review

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:518]: I have a slight preference in favor of `to_schnorr_keypair`. 

I also like `to_ecdsa_privkey`, `to_ecdsa_pubkey`, and `to_schnorr_pubkey` over the current ones. But this one is more debatable since the user has the context of what return type to expect.

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:641]: Let's stick with the term `X-only` instead of `x-coord-only` to be more consistent with bip340

**Review by sanket1729 — COMMENTED** [review]
Thanks for this. The code looks great. I have a discussion point about API naming which might use more feedback from other contributors.

File: src/util/bip32.rs
Comment (dr-orlovsky) [reply src/util/bip32.rs:518]: Changed to `to_keypair` since there is no other keypair (and @apoelstra does not like "Schnorr" surname :D because of him patenting signature algo)

File: src/util/bip32.rs
Comment (dr-orlovsky) [reply src/util/bip32.rs:641]: Last time I read it it was "X-coord only", but anyway, changed. Omnia mutantur...

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:643]: This returns a public key and not a kaypair

File: src/util/bip32.rs
Comment (sanket1729) [reply src/util/bip32.rs]: I would prefer `to_schnorr_keypair`. It's sad that this has patent history, but IMO the convenience and usability triumphs this.

Comment (dr-orlovsky) [reply src/util/bip32.rs]: Do we have a keypair type for non-BIP-340 keys?

Comment (dr-orlovsky) [reply src/util/bip32.rs]: I really do not like inconsistency in function names for extended keys that we have today in the API: `derive_priv`, `derive_pub`, `new_master`, `from_private` - and now `to_ecdsa_privkey`.... :(

Comment (dr-orlovsky) [reply src/util/bip32.rs]: Did a temporal version with `priv/pub` instead of `privkey` for more consisted naming for now; will wait for others to comment

File: src/util/bip32.rs
Comment (sgeisler) [inline src/util/bip32.rs:508]: I wonder if the `_priv` part is really useful. You know you are operating on an xpriv and the return type is a private key. Kinda related: I'd like to explore the possibility of putting the derive and conversion functions into a trait (not here though) to allow generic code over x-keys.

File: src/util/bip32.rs
Comment (sgeisler) [inline src/util/bip32.rs:518]: Same here `_keypair` seems redundant.

File: src/util/bip32.rs
Comment (sgeisler) [inline src/util/bip32.rs:634]: redundant `_pub`

File: src/util/bip32.rs
Comment (sgeisler) [inline src/util/bip32.rs:643]: redundant `_pub`

**Review by sgeisler — DISMISSED** [review]
Looks pretty good. I just now saw @sanket1729's comment :laughing: so take my comments with a grain of salt. As explained in the first comment my aversion to the more descriptive names comes from the idea of unifying the interface of x-keys via a common trait. For that the different names would be very annoying (but no blocker here, won't cut a release too soon anyway). I'll try to build it in a separate PR, the basic idea is to have:

```rust
pub trait XKey: Sized {
    type EcdsaKey;
    type SchnorrKey;

    fn derive<P: IntoIterator<Item=ChildNumber>>(&self, path: P) -> Self;
    fn to_ecdsa(&self) -> Self::EcdsaKey;
    fn to_schnorr(&self) -> Self::SchnorrKey;
}
```

**Comment by apoelstra** [comment]
What is the status of this? Does it need a rebase and/or the PR description updated to have different commit IDs?

**Comment by dr-orlovsky** [comment]
@apoelstra rebased, description update, ready for (re)reviews

File: src/util/bip32.rs
Comment (sanket1729) [reply src/util/bip32.rs]: Reviewing the discussion, after some thinking I think my vote is for `to_ec_pubkey`, `to_xonly_keypair` for rationale mentioned here: 
https://github.com/rust-bitcoin/rust-bitcoin/issues/588#issuecomment-890586744

**Review by sanket1729 — COMMENTED** [review]
I think we should resolve our discussion https://github.com/rust-bitcoin/rust-bitcoin/issues/588#issuecomment-890586744 before proceeding with this.

**Comment by dr-orlovsky** [comment]
@sanket1729 that discussion much more relates to previous PR #589 which is already merged (and a bitcoin-rust version with it is already released). If we need to change that approach, that should be a matter of a new PR.

This PR is very specifically removes unnecessary non-Secp256k1 pubkeys from BIP32/PSBT related data structures and does not directly related to the key type naming discussion within rust-bitcoin

**Comment by sanket1729** [comment]
@dr-orlovsky, there is a discussion to possibly undo things in #589 (in a non-breaking way). And there seems to be some support for that. If there is a consensus for that, this PR would be updated, namely, the function names that use `ecdsa_` and `schnorr_` should be `ec_` and `xonly_`(`x_only_`)

**Comment by dr-orlovsky** [comment]
Why we can't do that alltogether after merging this PR, independently from it?

**Comment by sanket1729** [comment]
I don't have a strong opinion but appears that there is consensus for removing signature names from functions names using `x_only`, `ec`. 

This PR is not dependant on it and can get in without the other one, but there is no reason to have poor API names here just so we can change them later.

**Comment by dr-orlovsky** [comment]
@sanket1729 can you pls explain what exactly you propose to update in this PR?

**Comment by sanket1729** [comment]
I was about to suggest APIs like 
```diff
- pub fn to_ecdsa_priv(&self) -> ecdsa::PrivateKey
+ pub fn to_ec_priv(&self) -> ecdsa::PrivateKey
```
but reading this seems odd that API when the return type is odd. 
ACK 7c565ec7a65e423756c42fc89b28a4fa7fccf80e.

We can change this later when we address #588

**Review by sanket1729 — DISMISSED** [review]
ACK 7c565ec . Once we rename `ecdsa` and `schnorr` keys to `ec` and `xonly`, we should update the API names in this PR too.

File: examples/bip32.rs
Comment (apoelstra) [inline examples/bip32.rs:61]: I'm a little tempted to just impl `From<secp256k1::PublicKey>` for `ecdsa::PublicKey` which does this.

File: src/util/bip32.rs
Comment (apoelstra) [inline src/util/bip32.rs:522]: Can you add a unit test which tries to parse an xpriv with a secret key of all zeros and one where the secret key is all `ff`s? I'm worried that this panic can be triggered.

**Comment by GeneFerneau** [comment]
utACK [7c565ec](https://github.com/rust-bitcoin/rust-bitcoin/pull/590/commits/7c565ec7a65e423756c42fc89b28a4fa7fccf80e)

File: examples/bip32.rs
Comment (thomaseizinger) [reply examples/bip32.rs:61]: There is already a `new` function that defaults to compressed IIRC.

File: src/util/bip32.rs
Comment (dr-orlovsky) [reply src/util/bip32.rs:522]: Well, you can't construct xpriv with such secret keys - it errors during the construction on incorrect secret key. However I added general test case for this method.

File: src/util/bip32.rs
Comment (Kixunil) [inline src/util/bip32.rs:882]: Is this used so often that removing three letters is worth readability and larger diff?

Comment (dr-orlovsky) [reply src/util/bip32.rs:882]: It was not removed for readability, but for standardization. Most of method names uses `pub`/`priv` or `secret`, not `private`

**Comment by dr-orlovsky** [comment]
@apoelstra I added the test in the last commit (by amending it). Unfortunately it required to change crate policy from `forbit(unsafe_code)` to `deny(unsafe_code)` but only in test configuration. The function does not panics but errors.

**Comment by apoelstra** [comment]
@dr-orlovsky that's cool, but I think you're misunderstanding the test that I want. What you've done (involving unsafe code) seems to just test rust-secp functions .... somewhich which not only belongs in rust-secp, but which doesn't require unsafety even if you were testing it here.

What I would like is a stringified xpriv with a bad privkey embedded in it, and an attempt to parse it into a xpriv struct.

**Comment by dr-orlovsky** [comment]
@apoelstra done

File: src/util/bip32.rs
Comment (apoelstra) [inline src/util/bip32.rs:1124]: In ca4e90cd3eae74ad6f7a93af22aacf3eec53e73d:

I have a mild preference for using `unwrap_err` (and then maybe `to_string`) to compare errors inside the test, rather than using `should_panic`.

File: src/util/bip32.rs
Comment (apoelstra) [inline src/util/bip32.rs:1156]: In ca4e90cd3eae74ad6f7a93af22aacf3eec53e73d:

Note to reviewers: this has a chaincode of all zeros but the secret key is all `ff`s.

**Review by apoelstra — APPROVED** [review]
ACK ca4e90cd3eae74ad6f7a93af22aacf3eec53e73d

File: src/util/bip32.rs
Comment (sanket1729) [reply src/util/bip32.rs:516]: After #755, I think this should be just `to_priv`. As mentioned and discussed in several places, keys are not related to signature algorithms.

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:630]: I am not a big fan of this change. But I can live with this. This basically brings a compile error downstream that is difficult to solve. I think we should at least add this in release notes. 

The user at least has some guidance from the compiler when we change the function signature while keeping the name the same. In this case, it would be just say function not found.

File: src/util/bip32.rs
Comment (sanket1729) [reply src/util/bip32.rs:642]: Same, this should be `to_pub`

File: src/util/bip32.rs
Comment (sanket1729) [reply src/util/bip32.rs:651]: And this should be `to_x_only_pub`

**Review by sanket1729 — COMMENTED** [review]
If we can rebase this after #755, it would be great. I think the naming can improve substantially then.

**Comment by dr-orlovsky** [comment]
@sanket1729 addressed your review and re-based. From what I see from the failed test in the last commit we need another point release for secp:

```
binary operation `==` cannot be applied to type `secp256k1::KeyPair`
```

I was working before with the previous secp version.

File: src/util/bip32.rs
Comment (Kixunil) [inline src/util/bip32.rs:653]: This method feels more like a getter because it loses some information (conversion like `to_` usually don't) . Getters in Rust don't have prefixes (`get_` or `to_`), so the name should be `public_key()` I think.

File: src/util/bip32.rs
Comment (Kixunil) [inline src/util/bip32.rs:659]: This conversion looks wrong. If I understand this correctly it may change the parity of the key making it effectively different than the original one. I can't imagine how it'd be a good idea to do that and even if it has some use cases, having it here publicly could give people wrong ideas. Please explain if I'm missing something.

FWIW I'm not even sure if `impl From<PublicKey> for XOnlyPublicKey` was a good idea given it loses one bit of information.

Also, same issue with getter as above.

**Review by Kixunil — CHANGES_REQUESTED** [review]
The strange conversion is too worrying to me to ACK this, please explain or remove.

(side note: abbreviating `private` to `priv` hurts my eyes but not going to argue about it)

**Comment by dr-orlovsky** [comment]
@Kixunil it is a part of the standard to drop first byte in public key - to x-only public key conversion for BIP32: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki#public-key-conversion

As regarding `priv` vs `private`, I changed the name to follow already used naming of other methods. I do not mind renaming all methods from `priv` to `private` but this will be a larger breaking change.

**Comment by apoelstra** [comment]
I think @dr-orlovsky is correct. I agree, it's weird to drop a bit of information in a "conversion" but this bit of information is not used in signature validation, and this is the standard way to obtain an x-only key from an ordinary one.

**Review by Kixunil — DISMISSED** [review]
ACK 1ff5730d21bad940d18b975199b2c0a393056644

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:1126]: I don't see any value in this test? What is this test testing? It is a one-line function code that is repeated in the test. 

If you feel there is value in this, we can check that the x-only public keys from the keypairs are the same for now. After another point release, we can re-add this test. I don't think we are losing any features for the next release because of this.

**Review by sanket1729 — COMMENTED** [review]
Left two comments. I don't think we should wait for a point release of secp for a purpose of a test (which I think is not testing anything concrete). We can add the test later after a point release of secp.

**Comment by dr-orlovsky** [comment]
@sanket1729 addressed issues and removed test.

@Kixunil pls re-ACK

**Comment by Kixunil** [comment]
Would've reacked but it's uncompilable...

**Comment by dr-orlovsky** [comment]
@Kixunil sorry, my bad: introduced last minute "improvement" with additional commit and didn't properly checked it. Should be fixed now

**Comment by Kixunil** [comment]
Still some issue.

**Comment by dr-orlovsky** [comment]
@Kixunil I do not know what the ... is happening with CI, but it perfectly builds locally while it fails with exactly the same config on GitHub:

```
orlovsky@alienware:~/repos/rust-bitcoin$ rustup update
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: syncing channel updates for 'nightly-x86_64-unknown-linux-gnu'
info: cleaning up downloads & tmp directories
orlovsky@alienware:~/repos/rust-bitcoin$ cargo +stable build --features=no-std --no-default-features
    Finished dev [unoptimized + debuginfo] target(s) in 0.27s
orlovsky@alienware:~/repos/rust-bitcoin$ cargo +nightly build --features=no-std --no-default-features
    Finished dev [unoptimized + debuginfo] target(s) in 0.26s
```

**Comment by sanket1729** [comment]
@dr-orlovsky , the tests are not building. Not the binary.

**Comment by sanket1729** [comment]
This probably has something to do with `DisplaySecret` PR in rust-secp256k1. I think you are probably the best person to debug this since you authored it :P

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:49]: Why are you doing this? I think this is causing the issue? Why cannot use derive debug in no-std? And how is that related to this PR.

**Comment by sanket1729** [comment]
I think we should leave all the secret display stuff out of this PR and this should build.

File: src/util/bip32.rs
Comment (dr-orlovsky) [reply src/util/bip32.rs:49]: We need this since `SecretKey` in the upstream does custom `Debug` which works only on `std`. Without this change it foes not compile on `no_std`

**Comment by dr-orlovsky** [comment]
@Kixunil @sanket1729 sorry it was my bad: probably was too tired yesterday and thought that I had no-std `Debug` impl for `ExtendedPrivateKey`. Fixed

**Review by apoelstra — APPROVED** [review]
ACK cf0c48cc86b1f9710588987db8137362275166ad

It's a little eyebrow-raising that the nostd `Debug` impl just silently omits the secret key, but whatever. I have no opinions about the renaming, though I think after we get all the key-changing PRs in (but before we do the Taproot release) we should go through everything and possibly do a rename cleanup.

**Comment by dr-orlovsky** [comment]
@apoelstra there is not a lot of options for `no_std` `Debug`: in `secp256k1` secret key do not have any `Debug` in `no_std` and is hashed in `std`. If we just plainly print secret key here we will abandon "never display secrets in logs" policy we introduced in `secp256k1` with that hashing.

**Comment by apoelstra** [comment]
Right, I understand that we have no good options here. I think I might've stuck the literal string `[secret]` in there or something, but just eliding the field is fine.

File: src/util/bip32.rs
Comment (sanket1729) [inline src/util/bip32.rs:664]: Note that there would be some conflicts with #757

**Review by sanket1729 — APPROVED** [review]
ACK cf0c48cc86b1f9710588987db8137362275166ad. #757 might need rework after this

**Pull request merged**
